Round3[Step2] - Jay

shin-jae-uk opened the pull request:
@Gundy93 

건디! 안녕하세요~
Round3 Step2 PR 드립니다.

### Step1 결과물을 바탕으로 추가 요구사항에 따라 코드를 추가/수정했습니다.
  - `CoffeeShop` 타입을 기존 구조체에서 클래스로 변경했습니다.
  - 주문한 `coffee`의 가격(`menu` 딕셔너리의 `value`)을 추출하기 위해 `if-let` 구문으로 옵셔널 바인딩을 사용했습니다.
  - 잔액과 가격을 비교하여 상황에 따른 출력문을 구분했습니다.
---
### 고민되었던 점
1. 구조체 `Person`의 `order` 메서드에서 `CoffeeShop`의 `make` 메서드를 호출했더니 아래와 같은 오류가 발생했습니다. 

   > `Cannot use mutating member on immutable value: 'coffeeShop' is a 'let' constant`


   구조체 내부의 메서드가 반환하는 값은 불변값(상수)로, 이 반환값을 타 구조체의 가변 메소드에서 호출하면 에러가 발생한다는 것을 알았습니다. 이에 `CoffeeShop`을 클래스로 타입을 변경했습니다.
메서드 선언에 `mutating` 키워드를 붙이면 될 것이라 생각했는데 불가하여 해결에 고민이 많았습니다🤔 사실 현재도 완전히 이해가 되는 것은 아닌 것 같아요. 
_값타입인 구조체의 메서드를 호출했으니 복사인데, 외부에서 값을 변경하거나 수정하는 것은 모순이 되어서 그런 것인지.. 음.._ 🤨

2. 옵셔널인 딕셔너리의 값을 호출하여 Int타입 변수와 어떻게 비교를 해야할지 고민했습니다.
여러 삽질 끝에.. `if-let` 구문 속에 조건문을 넣어본 것은 처음인데, 잘 처리된 것 같아 기쁩니다!

3. 요구사항은 아니지만, 만약 한번에 여러 메뉴를 동시에 주문할 수 있는 기능을 구현한다면 어떻게 해야할지 고민이 됩니다.
현재도 아이스로 변경 메뉴를 넣어놓았지만 쓸 수 없네요😮‍💨

shin-jae-uk committed:
Step2 - 인스턴스는 서로 소통한다

Gundy93 commented on CodeStarterCamp_Week3/CoffeeShop.swift:
```diff
@@ -5,18 +5,23 @@
-    var totalSales: Int = 0
-    var menu: [Coffee: Int] = [.espresso: 4000, .americano: 5000, .cafeLatte: 5500, .cafeMocha: 6000, .vanillaLatte: 6000, .changeToIced: 0]
+class CoffeeShop {
+    var name: String
```
`name` 프로퍼티를 `var`로 선언한 이유는 무엇인가요?

'상호명을 바꿀 수 있다' 이런 느낌인가요?

shin-jae-uk replied:
`coffeeShop`의 `name` 프로퍼티가 변경될 일은 굳이 없을 것 같습니다!
상수로 수정했습니다.

Gundy93 commented on CodeStarterCamp_Week3/CoffeeShop.swift:
```diff
@@ -5,18 +5,23 @@
     }
     
-    mutating func makeCoffee(_ coffee: Coffee) {
+    func make(_ coffee: Coffee, from name: String) {
```
물론 이 함수가 손님의 '이름'을 호출하는 것은 맞지만, 이 상황에서는 `name`보다는 `customer`와 같은 표현은 어떨까요?

`from`과의 호응도 그 편이 좋을 것 같다고 생각합니다. 꼭 그렇게 바꾸라는 것은 아닙니다. 😃

shin-jae-uk replied:
자연스러운 표현에 `customer`로 변경이 좋을 것 같습니다.

Gundy93 commented on CodeStarterCamp_Week3/Person.swift:
```diff
@@ -6,10 +6,23 @@
 
 struct Person {
-    let name: String
+    var name: String
```
여기 `name`도 `var`로 바뀌었는데, 어떤 이유에서 `var`를 선택하게 됐을까요?

shin-jae-uk replied:
`name`이 '실명이 아니라면 추후 변경해도 되지 않을까' 라는 생각에서 `var`로 선택했던 것 같아요😅
일단 명칭이 nickName은 아니니, 다시 `let`으로 변경했습니다.

Gundy93 commented on CodeStarterCamp_Week3/Person.swift:
```diff
@@ -6,10 +6,23 @@
-    func orderCoffee(_ coffee: Coffee) {
-        print("\(name)이(가) \(coffee.rawValue)를 주문합니다.")
+        
+    mutating func order(_ coffee: Coffee, of coffeeShop: CoffeeShop, by name: String) {
```
물론 이 부분이 기획서에 나온 그대로 작성됐다는 것은 압니다만, `Person` 타입은 `name` 프로퍼티를 갖고 있는데 굳이 매개변수로 전달받을 이유가 있을까요?

종종 개발자는 기획이 적절하지 않을 경우에 당면할 수 있습니다. 그런 경우에는 협의를 할 수도 있는 부분이라 코멘트 남겨봅니다. 😃

shin-jae-uk replied:
`order` 메서드에서 굳이 매개변수로 전달받을 이유는 없을 것 같아요 (`by name: String` 부분 삭제)

Gundy93 commented on CodeStarterCamp_Week3/Person.swift:
```diff
@@ -6,10 +6,23 @@
+    mutating func order(_ coffee: Coffee, of coffeeShop: CoffeeShop, by name: String) {
+        print("\(name)이(가) \(coffeeShop.name)에서 \(coffee.rawValue)를 주문합니다.")
+        
+        if let price = coffeeShop.menu[coffee] {
```
이것을 통해 메뉴가 있는지 여부와 `price`라는 상수로 옵셔널 바인딩하는 두 작업을 잘 처리하신 것 같네요. 좋습니다. 👍

Gundy93 commented on CodeStarterCamp_Week3/Person.swift:
```diff
@@ -6,10 +6,23 @@
+        if let price = coffeeShop.menu[coffee] {
+            if money >= price {
+                money -= price
+                coffeeShop.totalSales += price
```
이건 마치 주문한 고객이 직접 커피숍의 돈통을 만지는 행위로 보입니다.

다른 객체의 프로퍼티를 직접 수정하는 행위는 객체 지향적인 설계와는 거리가 멀어보입니다. 다만 메서드 호출은 괜찮습니다.

돈을 받을 수 있도록 `CoffeeShop`에 적절한 메서드를 만들고 호출해 돈을 전달하는 편이 좋을 것 같습니다.

shin-jae-uk replied:
돈을 받고 매출 출력은 `CoffeeShop` 객체내 `getMoney()` 메서드에서 수행할 수 있도록 수정했습니다.

Gundy93 commented on CodeStarterCamp_Week3/Person.swift:
```diff
@@ -6,10 +6,23 @@
+                money -= price
+                coffeeShop.totalSales += price
+                print("\(price)원 결제가 완료되었습니다. (\(name)님의 현재 잔액: \(money)원)")
+                coffeeShop.make(coffee, from: name)
```
꼭 그렇게 변경해야 한다는 것은 아니지만, 앞선 코멘트에 있는 내용으로 만약 고객에게 돈을 받는 메서드를 만든다면 `make`를 `Person`이 호출하지 않더라도 내부적으로 호출되게 만들 수도 있을 것 같네요.

shin-jae-uk replied:
`make` 메서드를 내부적으로 호출한다는 말씀은, `getMoney` 메서드 내부에서 `make` 메서드를 호출한다는 뜻이신가요?
코드를 수정하여 호출 구조를 변경해봤습니다:)

Gundy93 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -6,10 +6,10 @@
-let missKim = Person(name: "Ms.Kim", money: 10000)
-var yagombucks = CoffeeShop(barista: misterLee)
+var misterLee = Person(name: "Mr.Lee", money: 5000)
+var missKim = Person(name: "Ms.Kim", money: 12500)
```
`misterLee`, `missKim`이라는 이름의 인스턴스가 `var`라는 것은 추후에 다음과 같이 변경될 수도 있다는 가능성이 열려있습니다.

```swift
misterLee = Person(name: "Mr.Park", money: 200)
```

이 부분에 대해서는 어떻게 생각하시나요?

물론 구조체로 정의한 이상 `money` 프로퍼티의 값을 `mutating`하기 위해서는 `var`일 수밖에 없긴 합니다.

shin-jae-uk replied:
네 맞습니다. 추후에 이름이 변경되는 상황이 생길 수 있겠네요..
`name`이 의미하는 바가 실명이면 변경되면 안되겠지만 만약 닉네임이라면 변경되어도 괜찮을 것 같아요.
`money` 프로퍼티 값이 변경되므로 인스턴스를 `var` 로 생성했었는데, 이러한 경우엔 `Person`을 클래스로 변경하는게 적절할까요?😃

Gundy93 replied:
아! 이건 아예 다른 사람(별개의 인스턴스)으로 변경될 수 있다는 의미로 남긴 코멘트였습니다.

변수 이름이 특정 identity를 나타내는 네이밍이라면 상수여야 할 것 같고, 특정 역할을 의미한다면 변수일 수 있을 것 같습니다.

전자인 경우에는 클래스가 적절할 것 같고, 후자의 경우에는 구조체도 괜찮을 것 같아요.

전자의 예시로는 철수, 영희 등이 있겠고, 후자의 예시로는 바리스타, 선생님 등이 있겠습니다.

Gundy93 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -6,10 +6,10 @@
-var yagombucks = CoffeeShop(barista: misterLee)
+var misterLee = Person(name: "Mr.Lee", money: 5000)
+var missKim = Person(name: "Ms.Kim", money: 12500)
+var yagombucks = CoffeeShop(name: "야곰벅스", totalSales: 0, menu: [.espresso: 4000, .americano: 5000, .cafeLatte: 5500, .cafeMocha: 6000, .vanillaLatte: 6000, .changeToIced: 0], pickUpTable: [], barista: misterLee)
```
오히려 `yagombucks`는 `var`일 필요가 없을 것 같아서 변수로 선언하신 이유가 궁금합니다.

shin-jae-uk replied:
`yagombucks`는 `let`으로 선언하는게 맞을 것 같습니다.
Step1에서, `CoffeeShop`이 `struct`였을 때 `yagombucks`를 `var`로 선언했었는데 Step2에 와서 `class`로 타입 변경 후 인스턴스 선언을 적절하게 변경하지 않았던 것 같아요.

Gundy93 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -6,10 +6,10 @@
-yagombucks.makeCoffee(.espresso)
+missKim.order(.americano, of: yagombucks, by: missKim.name) // 주문 성공
+// misterLee.order(.cafeLatte, of: yagombucks, by: misterLee.name) // 주문 실패(잔액 부족)
+// print("픽업대 현황: \(yagombucks.pickUpTable.joined(separator: ", "))")
```
여전히 픽업대에 아메리카노가 남아있는 것을 보니, `Person`이 이 커피를 가져가도록 기능을 만들어도 재밌을 것 같네요. 😃

Gundy93 commented on CodeStarterCamp_Week3/CoffeeShop.swift:
```diff
@@ -5,18 +5,23 @@
+    var name: String
+    var totalSales: Int
+    var menu: [Coffee: Int] = [:]
     var pickUpTable: [String] = []
```
이니셜라이저에서 값을 전달받을텐데, 프로퍼티에 기본값이 있는 이유는 무엇인가요?

shin-jae-uk replied:
이니셜라이저 작성 후 프로퍼티 관련 내용을 수정하지 않았던 것 같아요.
반영하여 수정했습니다!

Gundy93 reviewed (commented):
제이님 이번 스텝도 잘 진행하신 것 같네요. 좋습니다. 👍

약간의 수정을 거치면 더욱 객체 지향적인 코드가 될 것 같습니다.

---

### 고민되었던 점

> 구조체 Person의 order 메서드에서 CoffeeShop의 make 메서드를 호출했더니 아래와 같은 오류가 발생했습니다.
>
>>Cannot use mutating member on immutable value: 'coffeeShop' is a 'let' constant
>
>구조체 내부의 메서드가 반환하는 값은 불변값(상수)로, 이 반환값을 타 구조체의 가변 메소드에서 호출하면 에러가 발생한다는 것을 알았습니다. 이에 CoffeeShop을 클래스로 타입을 변경했습니다.
메서드 선언에 mutating 키워드를 붙이면 될 것이라 생각했는데 불가하여 해결에 고민이 많았습니다🤔 사실 현재도 완전히 이해가 되는 것은 아닌 것 같아요.
값타입인 구조체의 메서드를 호출했으니 복사인데, 외부에서 값을 변경하거나 수정하는 것은 모순이 되어서 그런 것인지.. 음.. 🤨

매개변수로 전달되는 값들은 원본 그대로 전달되는 것이 아니라 카피본을 전달받게 됩니다. 그리고 그 카피본이 자동으로 `let`, 즉 상수로 선언되게 됩니다.

그래서 상수로 선언된 `coffeeShop` 전달인자를 수정할 수 없다는 것입니다.

> 그런데 클래스는 왜 수정이 가능한가?

참조타입인 전달인자의 카피본은 마찬가지로 카피된 `let`이긴 합니다. 다만 이 상수는 참조 인스턴스가 적재된 메모리의 주소만 가지고 있을 뿐입니다. 그래서 그 메모리 주소를 가리키는 상수가 아닌 주소에 위치한 실제 인스턴스를 수정하는 것이기 때문에 상수 선언과 무관하게 수정할 수 있는 것입니다.

이는 비단 함수나 메서드의 전달인자 뿐만이 아니라 전역, 혹은 지역 변수로 선언한 상수도 마찬가지입니다.

관련해서 몇 가지 정도 실험을 해 보시면 금방 이해할 수 있을 것 같습니다.

> 옵셔널인 딕셔너리의 값을 호출하여 Int타입 변수와 어떻게 비교를 해야할지 고민했습니다.
여러 삽질 끝에.. if-let 구문 속에 조건문을 넣어본 것은 처음인데, 잘 처리된 것 같아 기쁩니다!

이 부분은 적절하게 처리된 것 같습니다. 👍

> 요구사항은 아니지만, 만약 한번에 여러 메뉴를 동시에 주문할 수 있는 기능을 구현한다면 어떻게 해야할지 고민이 됩니다.
현재도 아이스로 변경 메뉴를 넣어놓았지만 쓸 수 없네요😮‍💨

```swift
class CoffeeShop {
    
    ...
    
    func takeOrder(_ order: Coffee..., from customer: String, payment: Int) {
        totalSales += payment
        
        for coffee in order {
            make(coffee, from: customer)
        }
    }
    
    ...
}

struct Person {
    
    ...
        
    mutating func order(_ coffee: Coffee..., of coffeeShop: CoffeeShop, by name: String) {
        print("\(name)이(가) \(coffeeShop.name)에서 \(coffee.map { $0.rawValue }.joined(separator: ", "))를 주문합니다.")
        
        let prices = coffee.map { coffeeShop.menu[$0] }
        
        if prices.contains(nil) {
            print("요청하신 메뉴에 없는 메뉴가 포함되어 있습니다.")
        } else {
            let totalPrice = prices.compactMap { $0 }.reduce(0, +)
            
            if money >= totalPrice {
                money -= totalPrice
                coffeeShop.takeOrder(coffee, from: name, payment: totalPrice)
                print("\(totalPrice)원 결제가 완료되었습니다. (\(name)님의 현재 잔액: \(money)원)")
            } else {
                print("잔액이 \(totalPrice - money)원만큼 부족합니다. (\(name)님의 현재 잔액: \(money)원")
            }
        }
    }
    
    ...
}
```

방법은 많을 것 같습니다만, 이런 식으로도 가능할 것 같습니다. 만드는 일과 주문을 받는 일을 분리하는 것이죠! 그럼 이제 하나만 만들 줄 아는 메서드도 `takeOrder`가 반복해서 다른 값을 넣으며 호출하면 되니까요.

꼭 이렇게 하실 필요는 없습니다. 다양한 방법을 고민해봅시다.

shin-jae-uk committed:
refactor: Apply code review feedbacks

- Correct naming(types, method, parameter)
- Select type('var' or 'let')
- Add a method(getMoney())
- Modify other details
